### DISCUSSION TITLE
Change Dropbehaviour in Classic Mode

### DISCUSSION THREAD

**Discussion by dredhorse** [Pryaxis/TShock#2091] [Ideas]
Hi,

I would like to get the same drops as in Expert Mode also in Classic Mode.

This could probably be coded but for this, you need to understand how the drop and mode system works.

Has somebody already looked into it? I already found some information in the code of the api but I'm not really into c#.

The reason behind this is that I have a server for my kids and I don't want to start with expert mode there. And cheating the stuff in all the time is also not that fun.

**Comment by Levardio** [comment]
What's about detecting when boss is low hp and enabling exprt and disabling it back when boss is dead?

**Comment by Quinci135** [comment]
https://github.com/Quinci135/NormalBossBags

**Reply by dredhorse** [reply]
That looks cool, so I would have to add all the bosses and the drops to it. Where can you find the information about bosses and drops?
Visual Studio Code does resolve the data for the bosses (I guess it comes from the included references) but not for the drops. (for whatever reason).
Also I can't really find the sourcecode location of the referenced classes / values. (Still learning c#)

**Reply by Quinci135** [reply]
Have to look at decompiled terraria source code for a lot of things
TShock/pyraxis has a tool for this https://github.com/Pryaxis/TerrariaSourceGetter or you could use something like dnspy/ilspy
Probably should have made it clearer, but the Terraria.NPC method npc.DropBossBags() works with all bosses except for queen slime and empress of light thus I handled those differently. As I've read from many places and what I've seen, terraria code is not considered very good. I have no idea where empress of light and queen slime drop their items in code so I've used a method that creates treasure bags.
All bosses should drop their loot in normal mode as if it were expert mode from what I've tested (exactly what you want I believe)

Some useful documentation
Tmodloader: Has lots of details for the several vanilla things they need
https://github.com/tModLoader/tModLoader/wiki#Vanilla-references
Tshock packet information (and their other readme pages though admittedly not too many)
https://tshock.readme.io/docs/getting-started
Tshock's GetDataHandlers on deserializing data
https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/GetDataHandlers.cs
Example plugin on reading a packet (C# is my first programming language and still don't know that much so my condition syntax isn't great but it works as intended)
https://github.com/Quinci135/CustomDeathReason/blob/master/CustomDeathReason/CustomDeath.cs
Other plugins as examples
https://github.com/Pryaxis/Plugins
*Added some notes to NormalBossBags describing what goes on

**Reply by dredhorse** [reply]
I will need to figure out how to get the sources as I normally run macOS, will try to get this done when I boot into windows again. 
Thanks for all the links, still the code documentation is lacking for a newbie in c#.
I did "code" a similar plugin for bukkit like you did, https://github.com/dredhorse/DeathTpPlus but that is Java.
I like a challenge so I will take a closer look into the material.

**Reply by Quinci135** [reply]
Sources? As in https://github.com/Pryaxis/TerrariaSourceGetter ?
It's built with dotnet core and framework.
So you can use (while at that directory)`dotnet ./TerrariaSourceGetter.dll` and it should start (I'm also on macos)
Then use linux or macos as the server platform and whichever version you choose (it will download the respective server versions & platforms) and let it run. The game logic is the same between platforms, just has some extra stuff to make it run via mono and fna. Terraria server is just a client without rendering code.

**Reply by dredhorse** [reply]
thanks for that information. I can now get it to run after installing some stuff but I get 
Process is terminating due to StackOverflowException.#########################################################  ]  98.37 %
[1]    9011 abort      dotnet ./TerrariaSourceGetter.dll

**Reply by dredhorse** [reply]
But I will try your plugin out, you should submit it to the Plugin Repo.
